### PR TITLE
HPCC-24908 Roxie splitters using wrong variable

### PR DESCRIPTION
**Pull request by richardkchapman**
Some splitter issues spotted when looking at code

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
Smoketest. I have not been able (or tried) to create an example which actually went wrong with prior code.

### CODE REVIEW

File: roxie/ccd/ccdserver.cpp [hunk 1]
```diff
@@ -9048,7 +9048,7 @@ class CRoxieServerThroughSpillActivity : public CRoxieServerActivity
         // MORE - yukky code (and slow). Could keep them heapsorted by idx or something
         // this is trying to determine whether any of the adaptors will in the future read a given record
         unsigned minIdx = (unsigned) -1;
-        for (unsigned i = 0; i < numOutputs; i++)
+        for (unsigned i = 0; i < numOriginalOutputs; i++)
         {
```
[the remark this answers is not in the document]
Comment (ghalliday) [reply roxie/ccd/ccdserver.cpp:9052]: I think this test is actually correct as it was.  I'm not convinced by the code in resetOutputsUsed() and noteOutputUsed().  I would have expected them to set used[i], but that has probably already happened.

File: roxie/ccd/ccdserver.cpp [hunk 2]
```diff
@@ -9058,7 +9058,7 @@ class CRoxieServerThroughSpillActivity : public CRoxieServerActivity
 
     inline bool isLastTailReader(unsigned exceptOid)
     {
-        for (unsigned i = 0; i < numOutputs; i++)
+        for (unsigned i = 0; i < numOriginalOutputs; i++)
```
Comment (ghalliday) [inline roxie/ccd/ccdserver.cpp:9061]: I think this is also correct before.

**Review by ghalliday — COMMENTED** [review]
@richardkchapman I'm not convinced the changes are correct....

File: roxie/ccd/ccdserver.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (richardkchapman) [reply roxie/ccd/ccdserver.cpp:9052]: used[i] is set in queryOutput

**Comment by richardkchapman** [comment]
The way I read the code, numOriginalOutputs says how big the array is, and numOutputs gets updated to indicate how many are actually in use (and therefore have to be waited for when tracking how many outputs have been stopped).

But the indexes passed in will always refer to the positions in the original array, so we need to scan the entire array in these loops.

I do think nextFreeOutput is almost certainly dodgy though!

**Comment by richardkchapman** [comment]
I don't think it should be possible for queryOutput to be passed -1, so I will delete that check and the nextFreeOuptut code (meaning it will assert if -1 is passed).

I still think the change is correct though...

**Comment by richardkchapman** [comment]
@ghalliday

**Pull request merged**